h/services/list_groups.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+from h import models
+from h.models import group
+
+
+class ListGroupsService(object):
+
+    """
+    A service for providing filtered lists of groups.
+
+    This service filters groups by user session, scope, etc.
+
+    ALl public methods return a list of relevant groups,
+    as dicts (see _group_model) for consumption by e.g. API services.
+    """
+
+    def __init__(self, session, request_authority):
+        """
+        Create a new list_groups service.
+
+        :param _session: the SQLAlchemy session object
+        :param _request_authority: the authority to use as a default
+        """
+        self._session = session
+        self.request_authority = request_authority
+
+    def _authority(self, user=None, authority=None):
+        """Determine which authority to use.
+
+           Determine the appropriate authority to use for querying groups.
+           User's authority will always supersede if present; otherwise provide
+           default value—request.authority—if no authority specified.
+        """
+
+        if user is not None:
+            return user.authority
+        return authority or self.request_authority
+
+    def all_groups(self, user=None, authority=None, document_uri=None):
+        """
+        Return a list of groups relevant to this session/profile (i.e. user).
+
+        Return a list of groups filtered on user, authority, document_uri.
+        Include all types of relevant groups (open and private).
+        """
+        open_groups = self.open_groups(user, authority, document_uri)
+        private_groups = self.private_groups(user)
+
+        return open_groups + private_groups
+
+    def open_groups(self, user=None, authority=None, document_uri=None):
```

Comment:
Right now this service has `open_groups` and `private_groups` methods and they're both public. This may change as we go, both in terms of naming and access.

Reply:
NB: `open_groups` encapsulates the entirety of the former `AuthorityGroupService`